modules/frontend/querysharding.go

```diff
@@ -181,6 +198,14 @@ func mergeResponses(ctx context.Context, rrs []RequestResponse) (*http.Response,
 		}
 	}
 
+	if combinedTrace != nil {
+		var err error
+		combinedTraceBytes, err = combinedTrace.Marshal()
+		if err != nil {
+			return nil, errors.Wrap(err, "error marshaling combined trace at query frontend")
+		}
+	}
+
 	if shardMissCount == len(rrs) {
```

Comment:
We can move this block above `if combinedTrace != nil`: if we exit the loop and no querier returned something, just exit immediately.